Streamline tree-selector menu

mathieu1 opened the pull request:
Streamlines the menu from #7667 with
- Shorter text, with icons.
- Main checkbox no longer opens the dropdown.
  Instead it just selects all/none.
- Fewer menu items but more actions possible by adding checkboxes
  in the menu
- Menu stays open until clicking outside.

Here's how it looks:
![tree-selector](https://cloud.githubusercontent.com/assets/7703352/6042002/ac0360f4-ac81-11e4-8de3-d52666f825d9.gif)

Of course, this needs some CSS work for it to look right, but let's see what you guys think about the mechanics first.
JS comments / CSS help and suggestions warmly welcome!

With more CSS tweaking, we could make that menu more compact so that this could be a dropup instead of a dropdown to avoid hiding the first items in the list (currently, the top of the menu is hidden behing Jupyter, see below)
![dropup](https://cloud.githubusercontent.com/assets/7703352/6042140/bec6bc62-ac82-11e4-9446-d70122b871fa.png)

mathieu1 committed:
streamline tree-selector menu using checkboxes

streamlines the menu from #7667 with
- Shorter text, with icons.
- Main checkbox no longer opens the dropdown.
  Instead it just selects all/none.
- Fewer menu items but more actions possible by adding checkboxes
  in the menu
- Menu stays open until clicking outside.

This obviously needs CSS tweaking

mathieu1 committed:
add badge to each menu item

this is a possible way to address #7628

mathieu1 commented:
One could easily address #7628 using badges... (will fix the UI quirks later if this stays)
![badges](https://cloud.githubusercontent.com/assets/7703352/6043453/8e29074a-ac8c-11e4-8fda-59c6736a143e.png)

dsblank commented:
If the counts could be right-justified on the menu row, that would look nice.

jdfreder commented:
@mathieu1 I like this idea a lot!  

Few of points:
- I don't think that we should make the dropdown a drop-up.  I don't think it's that big of deal if the first few rows are not visible because the dropdown blocks them.
- I feel a little uneasy about the "stickyness" of the menu - the dropdown not hiding when you select an item.
- I think the count is a good idea.  If we decide not for all the items, I think we should at least show it for the running notebooks.

jdfreder commented:
@ellisonbg design review.

jasongrout commented:
Wow, the demos above look really nice!

mathieu1 commented:
@jdfreder I don't have strong feelings regarding stickyness, but here are two good things about it:
- a priori it doesn't slow down common actions (like select all (running) notebooks and shutdown), and can significantly speed up rarer ones (like select idle notebooks).
- more importantly, when clicking on an "undetermined" checkbox, the user usually doesn't know the outcome of clicking (whether it's going to select all or none), and actually expects to be able to click twice in case he doesn't get the result he wants.

dsblank commented:
Another use for the counts is to actually count the items (rather than using it to select them). I have an extension allowing students to turn in their notebooks for homework, and I'd like to know how many have/have not quickly. (Not critical, of course, because I can pop over to a terminal... but this would be handy).

mathieu1 commented:
@dsblank Maybe selecting all notebooks is easy enough to see how many there are.

mathieu1 committed:
add disabling, align badges and improve look

- Disable checkboxes when there is nothing to select
- right-alignment of badges in the menu
- HTML and CSS improvements for better look (not final)

Internal changes:
- Avoid use of `stopPropagation()`!
  Instead, replace bootstap's dropdown hide/show logic
  with suitable sticky-dropdown behaviour.
- Avoid selecting the `..` parent folder

mathieu1 committed:
invoke css

dsblank commented:
Yes, that is all that I mean... I would like the counts for notebooks... I
like this, with the counts for notebooks +1 (and other counts, too).

On Thu, Feb 5, 2015 at 7:47 AM, Mathieu <notifications@github.com> wrote:

> @dsblank <https://github.com/dsblank> Maybe selecting all notebooks is
> easy enough to see how many there are.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/pull/7681#issuecomment-73040632>.
>

mathieu1 commented:
@dsblank Ok thanks, I thought you wanted another counter.

Rebased to include #7687, which solves refreshing issues.

Also improved the global experience:
- Disable checkboxes when there is nothing to select
- Avoid selecting the parent folder `..`
- right-alignment of badges in the menu
- HTML and CSS improvements for better look (still needs reworking)
Yet another screenshot:
![disabled](https://cloud.githubusercontent.com/assets/7703352/6061631/91093f80-ad49-11e4-9847-4c8f03dc95e0.png)

mathieu1 committed:
fix badge display on firefox

apparently, the use of `pull-right` requires the badges
to appear before the checkbox in order to display
correctly on firefox.

Carreau commented:
1 - Awesome increase in usability.
2 - Stil a big code change for 3.0, so would like more votes. 
3 - small request/question to others : badge in the 'select all'  to  show the total number of selected items.

mathieu1 committed:
add select-all badge

mathieu1 commented:
About @Carreau's 3rd suggestion, the difficult question is really about the positioning of the badge. I don't quite see where to put it so that it doesn't look ugly... Current look below:
![select-all-badge](https://cloud.githubusercontent.com/assets/7703352/6072279/fb8e56e6-ad9f-11e4-9349-e97fb0535b6b.gif)

(BTW, for some reason all my checkboxes turned grey (using Chrome) all of a sudden, *even on master*, and I haven't been able to figure out why (cleaning cache didn't help, nor rebooting). Does anyone else have the same issue?)

takluyver commented:
Would it look nicer as a plain number without the circle? Also, if we do go for that, we should certainly reserve enough width for at least two digits so the things to the right of it aren't moving as you select stuff?

As to the numbers in the dropdown: I would find it more natural if they were constant, showing the total number of folders, notebooks, running notebooks or files in the current view, rather than changing as things are selected.

Carreau commented:
> As to the numbers in the dropdown: I would find it more natural if they were constant, showing the total number of folders, notebooks, running notebooks or files in the current view, rather than changing as things are selected.

While I disagree, I think you need to know how many thing you have selected mostly I want to know if I have 1 or many thing selected when I want to take an action like deleting. 

> (BTW, for some reason all my checkboxes turned grey (using Chrome) all of a sudden

That would't surprise me from the chrome team to change the appearance of elements.

mathieu1 commented:
A selection of counter styles:
![badge-style](https://cloud.githubusercontent.com/assets/7703352/6079178/f5d7fbbe-ae01-11e4-896c-957b4931f689.png)

mathieu1 committed:
clean up CSS and firefox clicks

- `#button-select-all` toggles the all/none checkbox
  and the all/none checkbox now works on firefox
- static appearance of the `#button-select-all` button
  (even on counter change, up to two digits)
- clean up `tree.less`

mathieu1 committed:
invoke css

mathieu1 commented:
Cleaned up a few remaining quirks. Already got some testing on FF/chrome/safari.

- Ready for review / testing
- Awaiting decision(s) on the counters 
- Further suggestions still welcome

jdfreder commented:
Since it seems like we are still undecided on some of the things in this PR, I added it to this week's meeting agenda to decide what release we will target it to.

ellisonbg commented:
I am +1 on fixing the following in 3.1:

* select all/none behavior of main checkbox
* Styling the different menu items (with icons).

I am -1 on the complexity of the item count tags, the sticky menu and the
checkboxes in the menu. Feels like over design.

On Mon, Feb 9, 2015 at 12:17 PM, Jonathan Frederic <notifications@github.com
> wrote:

> Since it seems like we are still undecided on some of the things in this
> PR, I added it to this week's meeting agenda to decide what release we will
> target it to.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipython/pull/7681#issuecomment-73581749>.
>



-- 
Brian E. Granger
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

takluyver commented:
We discussed this in the dev meeting today, and more or less agreed with @ellisonbg 's comment above - we want to keep a couple of small enhancements, but we don't want to add lots of features to this until it's clear which ones people actually want.

mathieu1 committed:
keep only basic changes

Kept changes:
- fewer, shorter menu items (too cryptic?), with icons.
- main checkbox does not open menu, but selects all/none.
- total selection count next to main checkbox (unsure if this is to be
kept. It does address #7628).

Removed changes:
- sticky menu.
- checkboxes and counters in menu.

mathieu1 commented:
Updated according to @ellisonbg's comments. Specifically:

Kept changes:
- fewer, shorter menu items (too cryptic?), with icons.
- main checkbox does not open menu, but selects all/none.
- total selection count next to main checkbox (unsure if this is to be
kept. It does address #7628).

Removed changes:
- sticky menu.
- checkboxes and counters in menu.

mathieu1 committed:
add tooltips

jdfreder commented:
I think now that the behavior has changed minimally, this can be retargeted for 3.1.  I'm excited about these changes.

jdfreder commented on IPython/html/static/tree/js/notebooklist.js:
```diff
@@ -382,78 +386,26 @@ define([
+            set_state = set_state || (selection_type === "select-folders" && $(item).data('type') === 'directory');
+            set_state = set_state || (selection_type === "select-notebooks" && $(item).data('type') === 'notebook');
+            set_state = set_state || (selection_type === "select-running-notebooks" && $(item).data('type') === 'notebook' && that.sessions[$(item).data('path')] !== undefined);
+            set_state = set_state || (selection_type === "select-files" && $(item).data('type') === 'file');
```
I have a minor style request here.  These lines are a little long, partially because you keep fetching `$(item).data('type')`.  Could you put that in a variable, like `type` or `item_type`?

jdfreder commented on IPython/html/static/tree/js/notebooklist.js:
```diff
@@ -382,78 +386,26 @@ define([
+            set_state = set_state || (selection_type === "select-files" && $(item).data('type') === 'file');
+            if (set_state) {
+                $(item).find('input[type=checkbox]').prop('checked', state);
             }
```
I'm not sure that this behavior makes sense anymore.  Why not remove the `state` function argument, renamed `set_state` to `state`, and get rid of the conditional?

mathieu1 committed:
simpler select function

mathieu1 commented:
@jdfreder Thanks for your comments, I just addressed them. The second suggestion changes a bit the behavior : selecting "all notebooks" for exemple will now also deselect everything else (before, it was just adding all notebooks to the selection, but not changing the rest). The current behavior is exactly like Gmail's.

Reminder to self/all: if merged for 3.x, this also fixes #7737

jdfreder commented:
Travis failure was sporadic.  I restarted the test.  @Carreau @ellisonbg or @minrk want to review and merge this?  I'm okay with it in it's current state :+1:

ellisonbg commented:
It would be good to get some more detailed review, but overall I like the current state. Thanks for the work on this! The only part of it that I am 50/50 on is the overall selected count display. I can go either way on this depending on what other's think.

minrk commented:
Looks good to me. The only thing I might change is the padding around the number, so that it doesn't jump when it changes. But I think it's good enough to merge right now, and we can continue with smaller improvements later.

mathieu1 commented:
@minrk Do you mean adding even more space so that, say, 3-digit numbers would fit? Anyhow, merging and keeping smaller fixes for later is fine with me as well.

jdfreder commented:
Thanks @mathieu1

minrk commented:
@mathieu1 I would pick a balance - at least two digits shouldn't change width, but maybe three would be okay. It's a tradeoff of how much space it should take up. Alternately, it could be placed somewhere else, such that expanding doesn't push other elements around.
